src/helpers.js

```diff
@@ -36,13 +55,27 @@ function mergeHash(source, target, old, keepRemoved) {
 
     Object.keys(source).forEach(function (key) {
         if ( target[key] !== undefined ) {
-            if (typeof source[key] === 'object' && source[key].constructor !== Array) {
-                var nested = mergeHash( source[key], target[key], old[key], keepRemoved );
-                target[key] = nested.new;
-                old[key] = nested.old;
-            }
-            else {
-                target[key] = source[key];
+            if (typeof target[key] === 'object' && typeof target[key].msgstr !== 'undefined') {
```

Comment:
I would agree that the code is not very readable, but I'd have to disagree that what it's doing is complex. It's essentially just:

* Are we tracking paths?
 * yes, does the source contain path info?
    * yes, use the source's msgstr to set the target's msgstr
    * no, use source's value to set the target's msgstr
    * regardless, handle nested keys if they exist
 * no, handle old format as usual

One option would be to create two additional helpers to replace the conditionals and provide better readability, so it's more clear what is happening where:

```js
// function isPathObject () {...}
// function isNestedTranslation () {...}
// defined above^
if ( target[key] !== undefined ) {
    if ( isPathObject(target[key]) ) {
        // track paths
        if ( isPathObject(source[key]) ) {
            target[key].msgstr = source[key].msgstr;
        } else {
            target[key].msgstr = source[key];
        }

        if ( isNestedTranslation(source[key]) ) {
            var nested = mergeHash( source[key], target[key], old[key], keepRemoved );
            target[key] = nested.new;
            old[key] = nested.old;
        }
    } else {
        // don't track paths
        if ( isNestedTranslation(source[key]) ) {
            var nested = mergeHash( source[key], target[key], old[key], keepRemoved );
            target[key] = nested.new;
            old[key] = nested.old;
        }
        else {
            target[key] = source[key];
        }
    }
}
```

<hr/>

In addition, I could pull the handling of the new format out into a separate helper which gets called by `mergeHash`.

```js
// ... within mergeHash()
if ( target[key] !== undefined ) {
    if ( isPathObject(target[key]) ) {
        // track paths
        var hashWithPaths = mergeHashWithPaths( key, source, target, old, keepRemoved );
        target = hashWithPaths.target;
        old = hashWithPaths.old;
    }
    else {
        // don't track paths
        if ( isNestedTranslation(source[key]) ) {
            var nested = mergeHash( source[key], target[key], old[key], keepRemoved );
            target[key] = nested.new;
            old[key] = nested.old;
        }
        else {
            target[key] = source[key];
        }
    }
}
```

where `mergeHashWithPaths` is defined as:

```js
function mergeHashWithPaths(key, source, target, old, keepRemoved) {
    if ( isPathObject(source[key]) ) {
        target[key].msgstr = source[key].msgstr;
    }
    else {
        target[key].msgstr = source[key];
    }

    if ( isNestedTranslation(source[key]) ) {
        var nested = mergeHash( source[key], target[key], old[key], keepRemoved );
        target[key] = nested.new;
        old[key] = nested.old;
    }

    return {
        target,
        old
    }
}
```

Happy to do one or both of these, or try something else if you agree that it'll improve the readability/maintainability.